### PR TITLE
Drop method exported from WASM + removing leaks.

### LINKED ISSUE #6317: Expose `drop` API from the App in JS
### Discord username

_No response_

### How important is this feature to you?

5 – Lack of it blocks me from doing my work in Enso

### Describe the idea you'd like to see implemented.

This is needed for cloud integration. Cloud needs to drop the current instance of Enso (whole WASM thing) and load a new one when switching projects, as every project can be connected with a different GUI version. Thus, we need to expose a `drop` function in JS that will drop the whole app (and thus unregister all event handlers in JS). Basically, this should be something along the lines of `mem::drop(app)` in Rust. 

### Is your feature request related to a problem?

_No response_

### Screenshots, screencasts, mockups.

_No response_

### Would you be willing to help us implement this feature?

No

### PR DESCRIPTION
**Pull request by farmaazon**
### Pull Request Description

Fixes #6317 

The `drop` method is available in the WASM object. This can be tested by typing `ensoglApp.wasm.drop()` in the dev console - all objects should be removed and all connections closed.

### Important Notes

* This PR fixed serveral leaks by this occasion
* A new tool for tracking leaks was added to prelude's `debug` module.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: lib/rust/prelude/src/debug.rs [hunk 1]
```diff
@@ -126,3 +126,72 @@ impl Drop for TraceCopies {
...
+// === LeakDetector ===
+// ====================
+
+/// A module containing an utility for detecting leaks.
+pub mod leak_detector {
+    use crate::*;
+
+    thread_local! {
+        /// The structure mapping the existing tracking copies with [`Trace`] structure to their
+        /// creation backtraces. You can print
```
Comment (vitvakatu) [inline lib/rust/prelude/src/debug.rs]: sentence is not finished

**Comment by PabloBuchu** [comment]
Hej @farmaazon thank you for this. Unfortunately this solution requires wasm to be fully loaded and initialised which is not cloud case. 

Our current flow is:
We initialise ensogl App and pass instance reference to authentication module which takes care of setting user session. Here for now with newDashboard flag set to false you can run old IDE which calls `.run` which loads and initialise wasm.. 

New dashboard reinitialise ensoglApp on project run, sets `startup.project` and then calls `ensoglApp.run`. So unfortunately we don't have access to `ensoglApp.wasm`

Let me know what time suits you best tomorrow for a call so we can check whats are our options :)

**Comment by wdanilo** [comment]
But `ensoglApp.run` initializes WASM so ... we have access to what Adam exposed I think. Or I don't understand something. anyway, add me to the call tomorrow, please.

**Comment by vitvakatu** [comment]
@PabloBuchu @wdanilo fyi, Adam is off tomorrow.

**Comment by wdanilo** [comment]
@vitvakatu thanks a lot! In the meantime @PabloBuchu has written to me that this might be resolved. @PabloBuchu am I right here?

**Comment by sylwiabr** [comment]
can we merge this?who is doing QA here?

**Comment by PabloBuchu** [comment]
@sylwiabr @wdanilo 

QA 🟢 we should merge this so in cloud we could use the drop method to switch between project

Thank you for quickly delivering this :)

**Comment by somebody1234** [comment]
not sure if it's out of scope for this PR, but `.drop()` should probably remove the root element as well?

**Comment by farmaazon** [comment]
> can we merge this?who is doing QA here?

First, it should be reviewed. I've got no accept from anyone. @wdanilo @MichaelMauderer can you review it?

File: app/gui/src/lib.rs [hunk 2]
```diff
@@ -159,16 +177,37 @@ pub fn main() {
...
+/// All connections will be closed and all visuals will be removed.
+#[wasm_bindgen]
+pub fn drop() {
+    IDE.with(RefCell::take);
+    EXECUTOR.with(RefCell::take);
+    leak_detector::TRACKED_OBJECTS.with(|objects| {
+        let objects = objects.borrow();
+        if !objects.is_empty() {
+            error!("Tracked objects leaked after dropping entire application!");
+            error!("{objects:#?}");
```
Comment (MichaelMauderer) [inline app/gui/src/lib.rs]: Should this message be prefaced with some info? Maybe `error!("Objects that go leaked: {objects:#?}");`

**Comment by wdanilo** [comment]
I'll review it in 1h!

**Comment by vitvakatu** [comment]
> not sure if it's out of scope for this PR, but .drop() should probably remove the root element as well?

We're not creating it (it is hardcoded in the HTML), so I don't think we should remove it.

**Comment by somebody1234** [comment]
oh sorry brainfart, i meant the scene element - otherwise the new scene element is hidden under the old one so it just looks blank

File: lib/rust/prelude/src/debug.rs [hunk 3]
```diff
@@ -126,3 +126,75 @@ impl Drop for TraceCopies {
         }
     }
 }
+
+
+
+// ====================
```
Comment (MichaelMauderer) [inline lib/rust/prelude/src/debug.rs:142]: Should this be in the prelude, or is there a better place for it closer to the executor or other memory management functionality?

**Review by MichaelMauderer — APPROVED** [review]

File: lib/rust/prelude/src/debug.rs (hunk 3 above)
Comment (farmaazon) [reply lib/rust/prelude/src/debug.rs:142]: It needs to be available from many places, including both IDE controllers and view. Also, the executor I know is more about handling asynchronous tasks, not memory management.

I cannot think about any other package.

File: lib/rust/prelude/src/debug.rs [hunk 4]
```diff
@@ -126,3 +126,75 @@ impl Drop for TraceCopies {
...
+// === LeakDetector ===
+// ====================
+
+/// A module containing an utility for detecting leaks.
```
Comment (wdanilo) [inline lib/rust/prelude/src/debug.rs:146]: I was reading documentation of this module and the inner structs and I got the high-level overview, but there was no examples how this should be used and no explanation of how to use it. The constructor function documentation `Create enabled structure with appointed entity name (shared between all copies).` is not helpful either - I mean, it creates an "enabled" structure with a shared name, but then what? I think I'd need a little bit more docs here.

Also, I don't see how it's used in the new code, but that's probably it is somehow connected to the existing `TraceCopies` and I need some docs to understand it better.

**Review by wdanilo — APPROVED** [review]

File: lib/rust/prelude/src/debug.rs (hunk 4 above)
Comment (farmaazon) [reply lib/rust/prelude/src/debug.rs:146]: > Create enabled structure with appointed entity name (shared between all copies). is not helpful either - I mean, it creates an "enabled" structure with a shared name, but then what? I think I'd need a little bit more docs here.

Well, then it does exactly what struct's documentation says: 
* The underlying TraceCopies tracks copies, printing when they are created and dropped.
* Additionally, the existing copies are stored in [`TRACKED_OBJECTS`].
The documentation should be read with `TraceCopies` the `lead_detector::Trace` is a wrapper over it (and documentation says it).

What information do you still miss?

Comment (wdanilo) [reply lib/rust/prelude/src/debug.rs:146]: Ok, I think adding a clear and explicit information that all the `Traces` are stored in `TRACKED_OBJECTS` when the leak detector is enabled is good enough, but I missed that from docs. Maybe someone else would find it obvious, but I don't. I'd be very thankful for adding a line with this explanation there.

**Comment by vitvakatu** [comment]
## QA Report 🟡 

@somebody1234 noted that the scene element is left when we drop the application. We're creating a bunch of elements, so we should probably remove them. I don't know if it's necessary for cloud needs.

These elements are left after drop:

Screenshot 2023-04-24 at 23 19 05

Otherwise looks good.

**Comment by somebody1234** [comment]
it's easy enough to remove on the cloud side - however i think here is the right place to handle it

also note that all the other elements are children of the scene element so just removing the scene element should be enough

**Comment by PabloBuchu** [comment]
@farmaazon are you able to add removing `div.scene` element from DOM?

**Comment by farmaazon** [comment]
> @farmaazon are you able to add removing `div.scene` element from DOM?

Already there. Unfortunately, I spent some time on it trying to discover more leaks. Wasn't successful, but applied a workaround.

**Comment by vitvakatu** [comment]
## QA report 🟢 

Looks good to me, the only thing is that there are a bunch of warnings in the console:

image

But I guess this can be addressed later.

**Comment by somebody1234** [comment]
seems like ensogl is still running after the element is gone -
can confirm i have been experiencing that on the new dashboard where this is used

re: after the element is gone, the same error message (zero size) has happened before when i set the `display` to `none` accidentally (which hides the element, making `getBoundingClientRect()` return a rectangle with zero size). not sure if that helps though

**Pull request merged**